### ISSUE TITLE
Updating nodes and links without re-doing simulation

### ISSUE THREAD

**Issue by cobolfoo** [vasturiano/force-graph#154]
This is a way to update things like link color or node parameters without redoing the whole simulation? Let say I have one central node and 10 nodes connected to it. If I want to change one link color by using graphData() function, I get the whole simulation replays (the nodes start moving to place themselves). This is pretty annoying when you do it every second :+1:

**Comment by cobolfoo** [comment]
I was able to do it by hacking your code a bit. I created a method to expose the state variable from outside. Then I moved your paint functions into public scope. I manually update the state.graphData.nodes and state.graphData.links array. Then I call paint functions directly. Seem to work perfectly, if the number of nodes or links changes I perform a normal graphData() call.

**Comment by riccardoferretti** [comment]
Great timing, I was actually looking into a very similar issue. When I update the links/nodes, the whole simulation starts over (with all the problems you described above).
Ideally if I remove a link, I just want the link to go, and so on for the other operations.

@cobolfoo can you share more about your approach? maybe with some code?
So I can see whether it can be expanded to my use case.

**Comment by cobolfoo** [comment]
Since the original code use kapsule and need rolljs to compile a proper production javascript, here is the compiled version I use in my project. The only changes I made is that the ForceGraph object have a public getState() now and that the underlying canvas force graph objects also expose paint functions.
`
var state = this.graph.getState();
// from now on, state.graphData is editable (nodes and links)

// Then you update the visual
state.forceGraph.paintNodes();
state.forceGraph.paintLinks();
state.forceGraph.paintArrows();
state.forceGraph.paintPhotons();

state.shadowGraph.paintNodes();
state.shadowGraph.paintLinks();`


[modified_force-graph.zip](https://github.com/vasturiano/force-graph/files/5556267/modified_force-graph.zip)

**Comment by vasturiano** [comment]
You can cancel the re-heating of the force engine on updates by setting `.cooldownTicks(0)` before a data update. This will prevent the nodes from moving.

However, if you just need to modify a styling aspect of the nodes, you should not need to re-invoke `.graphData()`. You can simply set a new logic in `.nodeColor(node => /* your updated logic */)`. This should not cause the force engine to reheat.

The main reason why the force engine reheats by default at every invocation of `.graphData` is because if new data is passed onto the library, including potentially new or removed nodes, it is expected that the layout needs to react to this change in order to find a new balance.

**Comment by riccardoferretti** [comment]
thanks @vasturiano - I actually realized my issue is slightly different.

When I update my graph, I am recreating the whole data structure (`{nodes: [], links: []}`), so what I see is everything starting basically from scratch in terms of simulation.

I have played around, and I have noticed that if instead I mutate the objects in place, the simulation continues smoothly.

My guess is that there some problem around node identity, I thought it was driven by the `id` attribute in the node objects, but maybe it actually uses the object identity itself?
I saw that in d3 (I am new to it) there is this concept of Object Constancy and that [key functions](https://bost.ocks.org/mike/constancy/#key-functions) can be used to achieve it.

Am I doing something wrong? Do I need to manually compute the delta in the graph data, or can I use/pass a key function to achieve constancy?

Thanks!

**Comment by vasturiano** [comment]
@riccardoferretti yes you're right, the indexing of the nodes is done by object reference, not an object attribute like `id`.

Therefore to maintain continuity between data updates you should not create new node objects but instead mutate the existing ones, as you mentioned.

The reason why having an indexing key function is not possible is because `d3-force` does not make a defensive copy of the nodes data, but instead attaches position-related attributes to the nodes themselves. Thus, if the object is replaced by another one, all of that memory is gone and the node positioning starts from scratch.

You can read a bit more about it here: https://github.com/d3/d3-force#simulation_nodes

**Comment by riccardoferretti** [comment]
thanks very much for the explanation @vasturiano!
